File: src/18p2.py
with open("../inp/18tiny.txt", 'r') as f:
    lines = f.readlines()

r = 0
c = 0
points = [(r, c)]
cnt_boundary_points = 0
for line in lines:
    hex_value = line.split(" ")[2].strip("()#\n")
    d = hex_value[-1]
    cnt = int(hex_value[:-1], 16)
    if d == "0":
        c += int(cnt)
    elif d == "2":
        c -= int(cnt)
    elif d == "1":
        r += int(cnt)
    else:
        r -= int(cnt)
    points.append((r, c))
    cnt_boundary_points += int(cnt)

area = 0
for (p1, p0) in zip(points[1:], points[:-1]):
    area += (p0[0] + p1[0]) * (p0[1] - p1[1])
area = abs(area) // 2

print("Area (Shoelace formula) is", area)
print("Count of boundary points is", cnt_boundary_points)
cnt_interior_points = area - cnt_boundary_points // 2 + 1
print("Count of interior points (Pick's theorem) is", cnt_interior_points)
print("Interior + boundary is", cnt_interior_points + cnt_boundary_points)

File: src/18p2.kt
import kotlin.math.abs

fun main() {
    val lines = readInput("18")

    data class Point(val r: Long, val c: Long)

    var r = 0L
    var c = 0L
    var cntBoundaryPoints = 0L
    val points = mutableListOf<Point>()

    lines.forEach {
        val hexStr = it.split(" ").last().substring(2, 8)
        val d = hexStr.last()
        val cnt = hexStr.substring(0, 5).toLong(radix = 16)
        when (d) {
            '0' -> c += cnt
            '2' -> c -= cnt
            '1' -> r += cnt
            '3' -> r -= cnt
            else -> throw Exception("Unexpected direction $d")
        }
        points.add(Point(r, c))
        cntBoundaryPoints += cnt
    }

    val area = abs(points.zipWithNext().sumOf { (it.first.r + it.second.r) * (it.first.c - it.second.c) }) / 2
    val cntInteriorPoints = area - cntBoundaryPoints / 2 + 1
    (cntBoundaryPoints + cntInteriorPoints).println()
}